official/utils/arg_parsers/base.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Base parser class for all official model parsers."""
+
+import argparse
+import sys
+
+import tensorflow as tf
+
+
+# Map string to (TensorFlow dtype, default loss scale)
+DTYPE_MAP = {
+    "fp16": (tf.float16, 128),
+    "fp32": (tf.float32, 1),
+}
+
+
+class Parser(argparse.ArgumentParser):
+  """Base class for all official model parsers.
+
+  Implements the following features:
+    1)  --helpful for detailed command line flag help
+    2)  Enables secondary parsing for additional logic
+  """
+
+  def __init__(self, parents=None):  # pylint: disable=unused-argument
+    # this forces --helpful below --help in the flag summary
+    preemptive_parser = argparse.ArgumentParser(add_help=False)
+    preemptive_parser.add_argument("--helpful", "-H", action="store_true",
+                                   help="Detailed list of arguments.")
+    parents = [preemptive_parser] +(parents or [])
+
+    self.verbose_flags = self.aggregate_verbose_flags(parents=parents)
+    super(Parser, self).__init__(add_help=True, parents=parents)
+
+  def aggregate_verbose_flags(self, parents):
+    output = []
+    for parent in parents:
+      if "verbose_flags" in vars(parent):
```

Comment:
accesses a dict of the underlying variables. doing something like `vars(my_obj).get("foo")` is safer than `my_obj.foo` because it doesn't fail if `my_obj` doesn't have a `.foo`.